### ISSUE TITLE
False positive when using ^ with /g and lastIndex > 0

### ISSUE THREAD

**Issue by Exter-N** [uhop/node-re2#29]
While using `RegExp`, `^` only matches at the beginning of the string :
```js
let re = /^foo/g;
re.exec("foofoo"); // [ 'foo', index: 0, input: 'foofoo' ]
re.exec("foofoo"); // null
```

While using `RE2`, `^` matches at `.lastIndex` :
```js
let re = RE2("^foo", "g");
re.exec("foofoo"); // [ 'foo', index: 0, input: 'foofoo' ]
re.exec("foofoo"); // [ 'foo', index: 3, input: 'foofoo' ]
re.exec("foofoo"); // null
```

The problem seems to come from an optimization at https://github.com/uhop/node-re2/blob/7dc2814af81a2cffbeb3aed7707cd689f811f484/lib/exec.cc#L45-L55 which only takes the substring starting at `.lastIndex`, making `^` match at the relocated starting position.

**Comment by Exter-N** [comment]
I see three ways of fixing this : inspecting the pattern to see whether the optimization can be applied or not (can become quite tricky with `^`/`\A`, `\b`, `\B`, and maybe others), documenting these tokens as explicitly unsupported with `/g`, or unconditionnally undoing it.

**Comment by Exter-N** [comment]
This one should be fixed by #33.

**Comment by uhop** [comment]
I am pulling the optimization feature out because it should be an underlying library feature. It is impossible to detect an opportunity for it without proper parsing a regular expression and analyzing the resulting program.

Examples that break the fix now:

* `/(?:^a)/g`
* `/b|^a/g`

&hellip; and countless variations of those techniques. Just try it with something like `'aabc'`.

In general, I think it was a bad idea on my part. Right now I am working on removing it.
